File: haskell/BinarySearch.hs
-- Binary Search in Haskell (Functional Paradigm)
module BinarySearch where

-- Pure functional binary search implementation
-- Uses recursion and pattern matching
binarySearch :: Ord a => [a] -> a -> Maybe Int
binarySearch arr target = search 0 (length arr - 1)
  where
    search :: Ord a => Int -> Int -> Maybe Int
    search left right
      | left > right = Nothing
      | mid == target = Just midIndex
      | target < mid = search left (midIndex - 1)
      | otherwise    = search (midIndex + 1) right
      where
        midIndex = (left + right) `div` 2
        mid = arr !! midIndex

-- Phone book search example
type Contact = (String, String)  -- (Name, Phone Number)

phoneBookSearch :: [Contact] -> String -> Maybe Contact
phoneBookSearch contacts name = 
  let sortedContacts = sortBy (\(a,_) (b,_) -> compare a b) contacts
  in case binarySearch (map fst sortedContacts) name of
       Just index -> Just (sortedContacts !! index)
       Nothing   -> Nothing

-- Example usage
contacts :: [Contact]
contacts = 
  [ ("Om", "123-456-7890")
  , ("Prachi", "234-567-8901")
  , ("Deadhack", "345-678-9012")
  , ("Roshan", "456-789-0123")
  , ("Roshni", "567-890-1234")
  ]

-- Demonstration
main :: IO ()
main = do
  let result = phoneBookSearch contacts "Deadhack"
  case result of
    Just (name, phone) -> putStrLn $ "Found: " ++ name ++ " - " ++ phone
    Nothing -> putStrLn "Contact not found"

File: dart/binarySearch.dart
// Binary Search in Dart (Multi-Paradigm Approach)
class Contact {
  final String name;
  final String phoneNumber;

  Contact({required this.name, required this.phoneNumber});
}

class PhoneBookSearch {
  // Binary search implementation using an iterative approach
  static int? binarySearch<T extends Comparable<T>>(
      List<T> arr, T target) {
    int left = 0;
    int right = arr.length - 1;

    while (left <= right) {
      int mid = (left + right) ~/ 2;
      int comparison = arr[mid].compareTo(target);

      if (comparison == 0) {
        return mid;
      } else if (comparison < 0) {
        left = mid + 1;
      } else {
        right = mid - 1;
      }
    }

    return null;
  }

  // Phone book search method
  static Contact? searchContact(List<Contact> contacts, String name) {
    // Sort contacts by name
    contacts.sort((a, b) => a.name.compareTo(b.name));

    // Extract names for binary search
    List<String> names = contacts.map((contact) => contact.name).toList();

    // Perform binary search
    int? index = binarySearch(names, name);

    // Return contact if found
    return index != null ? contacts[index] : null;
  }

  // Main method to demonstrate usage
  static void main() {
    List<Contact> contacts = [
      Contact(name: 'Alice', phoneNumber: '123-456-7890'),
      Contact(name: 'Bob', phoneNumber: '234-567-8901'),
      Contact(name: 'Charlie', phoneNumber: '345-678-9012'),
      Contact(name: 'David', phoneNumber: '456-789-0123'),
      Contact(name: 'Eve', phoneNumber: '567-890-1234'),
    ];

    String searchName = 'Charlie';
    Contact? result = searchContact(contacts, searchName);

    if (result != null) {
      print('Found: ${result.name} - ${result.phoneNumber}');
    } else {
      print('Contact not found');
    }
  }
}

void main() {
  PhoneBookSearch.main();
}